virt_tests/strategy.py

```diff
@@ -0,0 +1,490 @@
+        if not snapsets:
+            return False
+
+        try:
+            data = json.loads(snapsets)
+            snapset_exists = any(
+                s["snapset_name"] == self.snapset_name for s in data.get("Snapsets", [])
+            )
+            return snapset_exists
+        except json.JSONDecodeError:
+            return False
+
+    def test_40_apply_system_changes(self) -> bool:
+        """Apply system changes that we want to test rollback from"""
+
+        # Install test packages
+        test_packages = " ".join(self.get_test_packages())
+        if not self.vm.run_command(
+            f"{self.get_package_manager()} update -y {test_packages}",
+            timeout=1200,
+        ):
+            return False
+
+        # Modify test files
+        commands = [
+            "echo 'modified-content' > /opt/test-baseline/test-file.txt",
+            "echo 'config-after-update' > /etc/test-config.conf",
+            "echo 'new-file-after-update' > /opt/test-baseline/new-file.txt",
+        ]
+
+        return all(self.vm.run_command(cmd) for cmd in commands)
+
+    def test_50_verify_changes_applied(self) -> bool:
+        """Verify that the system changes were actually applied"""
+
+        # Check file modifications
+        content = self.vm.get_command_output("cat /opt/test-baseline/test-file.txt")
+        if not content or content.strip() != "modified-content":
+            return False
+
+        config = self.vm.get_command_output("cat /etc/test-config.conf")
+        if not config or config.strip() != "config-after-update":
+            return False
+
+        # Check new file exists
+        if not self.vm.get_command_output("ls /opt/test-baseline/new-file.txt"):
+            return False
+
+        return True
+
+    def test_60_reboot_to_snapshot(self) -> bool:
+        """Reboot into snapshot entry to verify snapshot state"""
+
+        # Get snapset info to find snapshot entry
+        snapset_info = self.vm.get_command_output(
+            f"snapm snapset show {self.snapset_name} --json"
+        )
+        if not snapset_info:
+            return False
+
+        try:
+            data = json.loads(snapset_info)
+            if not data or not isinstance(data, list) or len(data) == 0:
+                return False
+
+            snapset = data[0]
+            snapshot_entry = snapset.get("BootEntries", {}).get("SnapshotEntry")
+
+            if not snapshot_entry:
+                err_print("No snapshot entry found")
+                return False
+
+            # Get boot title from boom
+            boot_title = self.vm.get_command_output(
+                f"boom list --noheadings -otitle --boot-id {snapshot_entry}"
+            )
+            if not boot_title:
+                err_print(f"Could not get boot title for entry {snapshot_entry}")
+                return False
+
+            log_print(f"⏩ Rebooting to snapshot entry: {boot_title}")
+
+            # Set next boot to snapshot entry
```

Comment:
@coderabbitai I don't agree with your nitpick point on this line regarding quoting `'` characters. From bash(1):

"Enclosing characters in double quotes preserves the literal value of all characters within the quotes, with the exception of $, `, \, and, when history expansion is enabled, !. ... The backslash retains its special meaning only when followed by one of the following characters: $, `, ", \, or <newline>."

Further:

"Enclosing characters in single quotes preserves the literal value of each character within the quotes.  A single quote may not occur between single quotes, even when  preceded  by  a backslash."

So your proposed quoting violates the shell rules for quoted strings in any case. Since the title should never contain `'` we shout just check for it and fail.